calyx-opt/src/passes/well_formed.rs

```diff
@@ -80,12 +80,71 @@ pub struct WellFormed {
     used_groups: HashSet<ir::Id>,
     /// Names of combinational groups used in the control.
     used_comb_groups: HashSet<ir::Id>,
-    /// ref cell types of components used in the control.
-    ref_cell_types: HashMap<ir::Id, LinkedHashMap<ir::Id, CellType>>,
+    /// ref cells of components used in the control. Used for type checking.
+    ref_cells: HashMap<ir::Id, LinkedHashMap<ir::Id, Cell>>,
     /// Stack of currently active combinational groups
     active_comb: ActiveAssignments,
 }
 
+enum Invoke<'a> {
+    StaticInvoke(&'a ir::StaticInvoke),
+    Invoke(&'a ir::Invoke),
+}
+
+impl Invoke<'_> {
+    fn get_ref_cells(&self) -> &Vec<(ir::Id, ir::RRC<Cell>)> {
+        match self {
+            Invoke::StaticInvoke(s) => &s.ref_cells,
+            Invoke::Invoke(s) => &s.ref_cells,
+        }
+    }
+
+    fn get_attributes(&self) -> &ir::Attributes {
+        match self {
+            Invoke::StaticInvoke(s) => s.get_attributes(),
+            Invoke::Invoke(s) => s.get_attributes(),
+        }
+    }
+}
+
+fn require_subtype(
+    invoke: Invoke,
+    self_ref_cells: &HashMap<ir::Id, LinkedHashMap<ir::Id, Cell>>,
+    id: &ir::Id,
+) -> CalyxResult<()> {
+    let cell_map = &self_ref_cells[id];
+    let mut mentioned_cells = HashSet::new();
+    for (outcell, incell) in invoke.get_ref_cells().iter() {
+        if let Some(oc) = cell_map.get(outcell) {
+            if !subtype(oc, &incell.borrow()) {
+                return Err(Error::malformed_control(format!(
+                    "The type passed in `{}` is not a subtype of the expected type `{}`.",
+                    incell.borrow().prototype.surface_name().unwrap(),
+                    oc.prototype.surface_name().unwrap()
+                ))
+                .with_pos(invoke.get_attributes()));
+            } else {
+                mentioned_cells.insert(outcell);
+            }
+        } else {
+            return Err(Error::malformed_control(format!(
+                "{} does not have ref cell named {}",
+                id, outcell
+            )));
+        }
+    }
+    for id in cell_map.keys() {
+        if !mentioned_cells.contains(id) {
+            return Err(Error::malformed_control(format!(
+                "unmentioned ref cell: {}",
+                id
+            ))
+            .with_pos(invoke.get_attributes()));
+        }
+    }
+    Ok(())
+}
+
```

Comment:
@sampsyo Was wondering if this looked ok w.r.t deduplicating some of the subtype code? It is used [here](https://github.com/calyxir/calyx/pull/2085/files#diff-4c8e5127c997c4a7aac5a0dbd742cde735638c78e2709a036ba0e8c09cefd4d8R631).